### PR TITLE
Persist last selected database in the native query editor for a user

### LINKED ISSUE #8780: Make new native queries default to using the last-selected database
Opening this on behalf of a user who [proposed this](https://twitter.com/dotpem/status/1055586084246056960). I think the behavior would simply be that when you click on new question -> native/sql, that it would default to whichever database you last selected in that database dropdown. This would remove a click if you tend to query the same database most of the time. 

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### LINKED ISSUE #39035: [Testing Plan] Make new native queries default to using the last-selected database
For issue: #8780

## Dimensions
- User
    - Id (this should be a user-local setting) 
    - Permissions
        - With native write permissions
        - Without native write permissions
    - Session (should persist after logging out and logging in again)
- Database type
    - Requires an additional table to run native query
    - Runs native query without a table
- Where native query editor is instantiated from
    - SQL query
    - Model
    - Action
- Does database still exist?
    - Yes
    - No

~~- Is there a query, was query run and was it saved?~~
    - This bit is irrelevant because the action is fired upon selecting a database and has nothing to do with actually running a query

## Actions
- Starting new (SQL Question | Model | Action)
- Database gets deleted after it's persisted
- User logs out, logs in again
- User logs out, some other user logs in in the same browser
- ~~Editing previously saved (SQL Question | Model | Action)~~

### Unit tests
_Will implement only backend unit tests._
- Make sure the setting can be set
- Make sure the setting can be read
- Make sure the setting is user-local only

### Integration tests
_Not going to implement these. They will require insane effort and will be almost as close as an E2E test._

### E2E tests
(Prerequisites: _set up postgres and mongo as additional databases, have at least one additional user in All users group_)
- As admin, open new native question
    - Choose a database (let's say Sample Database)
    - Navigate away from that page
    - Open new native question again
    - Make sure previously selected database is pre-selected
    - Log out
    - Log in as normal user
    - Open new native question
    - Make sure you are offered a choice of a database selection
    - Choose a different database (Mongo, for example)
    - Navigate away
    - Start new native question again
    - Make sure Mongo is the pre-selected database, and make sure the table picker is rendered
    - Log out
    - Log in as admin again
    - Start new native question
    - Make sure Sample Database is still selected for you, and not Mongo

- As admin, open new native question
    - Choose a database (let's say Postgres)
    - Go to admin settings and delete Postgres
    - Start new native question again
    - Make sure nothing breaks, and the UI prompts you to select database (Sample or Mongo)

- Make sure no redundant requests are fired when selecting the same database again

- Make sure this setting does not apply to model actions
    - Persisting a database for a native model or a native question should not influence model actions, and vice versa

- Make sure users without sufficient permissions cannot choose or persist databases they cannot query against

### PR DESCRIPTION
**Pull request by nemanjaglumac**
## Info
|||
|-|-|
| 📝 | [Product Doc](https://www.notion.so/metabase/Default-to-the-last-selected-database-in-native-queries-3bf8f844a6e8407a998bbf11a04f106b) |
| 🧪 | [Testing Plan](https://github.com/metabase/metabase/issues/39035) |

## What does PR accomplish?
This PR resolves #8780 by introducing a user-level setting for remembering the last used database in the native query.
Closes #39035.

```[tasklist]
### Tasks
- [x] Add new user-level setting to the backend
- [x] Wire native query editor to remember the last used database
- [x] Narrow the scope to just native questions and native models
- [x] Write unit tests
- [x] Write E2E tests
```

### About the `last-used-database` setting
- It should be user-local only
- It should work for any authenticated user; frontend ensures that a user without sufficient native permissions is not even offered to select and persist the last used database
- The information about the last used database currently should not be included in the audit information

> [!Note]
> The backend currently does have any additional checks to make sure the previously used database is still present. Frontend simply falls back to no database selected and prompts the user to select a new one.

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 1]
```diff
@@ -909,6 +921,14 @@ const mapDispatchToProps = (dispatch: Dispatch) => ({
     );
     return Questions.HACK_getObjectFromAction(action);
   },
+  onSetDatabaseId: (id: DatabaseId) => {
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:924]: Please remove `shouldPersistLastUsedDatabase`, add a callback to `NativeQueryEditor` like `onSetDatabaseId` and put the action in `View` and not here

**Review by ranquild — COMMENTED** [review]
@nemanjaglumac we would also need to handle database ids after the database is removed. I think we should do in clojure in `last-used-database-id` getter and return `null` if the corresponding DB cannot be found (could be done in another PR)

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx (hunk 1 above)
Comment (nemanjaglumac) [reply frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:924]: Help me understand this better.
What are we trying to achieve by moving the dispatch up to the `View`?

The reason I chose this approach was because I wanted a fine-grained control of when exactly do we persist the database. NativeQueryEditor can be rendered by many different components.

How are we going to control when does native query editor persists the database ID, and when does it ignore it?

Comment (nemanjaglumac) [reply frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx]: To answer my own question - this should do it.
https://github.com/metabase/metabase/pull/38986/commits/4dd6aad70815b5780baf5a675b2f62ed98146ebc

I'd still like to learn why is one approach considered better than the other?

File: e2e/test/scenarios/native/native-database-source.cy.spec.js [hunk 2]
```diff
@@ -0,0 +1,247 @@
+import { SAMPLE_DB_ID } from "e2e/support/cypress_data";
+import {
+  restore,
+  popover,
+  addPostgresDatabase,
+  POPOVER_ELEMENT,
+} from "e2e/support/helpers";
+
+const PG_DB_ID = 2;
```
Comment (kamilmielnik) [inline e2e/test/scenarios/native/native-database-source.cy.spec.js:9]: Isn't this `WRITABLE_DB_ID` ?

File: e2e/test/scenarios/native/native-database-source.cy.spec.js [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    startNativeQuestion();
+
+    assertSelectedDatabase(mongoName);
+    cy.findByTestId("native-query-top-bar").should("contain", "Select a table");
+  });
+});
+
+function startNativeQuestion() {
+  cy.visit("/");
```
Comment (kamilmielnik) [inline e2e/test/scenarios/native/native-database-source.cy.spec.js:203]: This is expensive. Can we use SPA navigation instead?

```suggestion
  cy.findByTestId("main-logo-link").click();
```

----

The same comment applies to other `cy.visit` calls in this test suite.

File: e2e/test/scenarios/native/native-database-source.cy.spec.js [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+      selectDatabase(additionalPG);
+
+      cy.log("Delete previously persisted database.");
+      cy.get("@postgresID").then(databaseId => {
+        cy.request("DELETE", `/api/database/${databaseId}`);
+      });
+
+      startNativeQuestion();
+      assertNoDatabaseSelected();
```
Comment (kamilmielnik) [inline e2e/test/scenarios/native/native-database-source.cy.spec.js:70]: I wonder if this test is going to pass if `startNativeQuestion` does not call `cy.visit`.
It should not, and we should ensure that the setting value is not stale (i.e. is refetched) when visiting native query editor.

File: e2e/test/scenarios/native/native-database-source.cy.spec.js [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+      );
+      selectDatabase("QA Postgres12");
+      cy.wait("@persistDatabase");
+
+      startNewAction();
+      assertNoDatabaseSelected();
+    });
+
+    describe("permissions", () => {
+      it("users with 'No self-service' data permissions should be able to choose only the databases they can query against", () => {
```
Comment (kamilmielnik) [inline e2e/test/scenarios/native/native-database-source.cy.spec.js:132]: I think this scenario (similar to when DB is removed) is also worth testing:

1. User has access to DB X
2. User creates a native query based on DB X
3. User is revoked permissions to DB X
4. User opens native query editor

Assertion: DB X is not listed and not selected

File: src/metabase/models/user.clj [hunk 6]
```diff
@@ -439,6 +439,12 @@
   :user-local :only
   :type :string)
 
+(defsetting last-used-database-id
+  (deferred-tru "The last database a user has used for a native query.")
```
Comment (kamilmielnik) [inline src/metabase/models/user.clj:443]: I suggest to be more specific with the name.

```suggestion
(defsetting last-used-native-query-database-id
  (deferred-tru "The last database a user has used for a native query.")
```

**Review by kamilmielnik — COMMENTED** [review]
Beautifully written test suite <3

File: frontend/src/metabase/query_builder/components/DataSelector/TriggerComponents.tsx [hunk 7]
```diff
@@ -80,7 +80,12 @@ export function FieldTrigger({
 
 export function DatabaseTrigger({ database }: { database: Database }) {
   return database ? (
-    <span className="text-wrap text-grey no-decoration">{database.name}</span>
+    <span
+      className="text-wrap text-grey no-decoration"
+      data-testid="selected-database"
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/query_builder/components/DataSelector/TriggerComponents.tsx:85]: Yes, it is the same because they can never be rendered at the same time. It's either-or type of situation, but the meaning of that element is the same - it shows the selected database.

1. when it's the only option and thus rendered by default
2. when it's the chosen option out of many

File: frontend/src/metabase/query_builder/components/view/View.jsx [hunk 8]
```diff
@@ -397,4 +403,11 @@ class View extends Component {
...
 
-export default ExplicitSize({ refreshMode: "debounceLeading" })(View);
+const mapDispatchToProps = dispatch => ({
+  onSetDatabaseId: id => dispatch(rememberLastUsedDatabase(id)),
+});
+
+export default _.compose(
+  ExplicitSize({ refreshMode: "debounceLeading" }),
+  connect(null, mapDispatchToProps),
+)(View);
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/query_builder/components/view/View.jsx:413]: It's funny cause [I used that exact same logic originally](https://github.com/metabase/metabase/pull/38986/commits/bffccafc66dd8b30707ce5f4724512e9d1e35bd8) (defined the action in the `NativeQueryEditor`) but @ranquild [asked me](https://github.com/metabase/metabase/pull/38986#discussion_r1499174779) to lift it to the `View`.

File: e2e/test/scenarios/native/native-database-source.cy.spec.js (hunk 2 above)
Comment (nemanjaglumac) [reply e2e/test/scenarios/native/native-database-source.cy.spec.js:9]: In theory yes.
But it can be misleading. We're using `WRITABLE_DB_ID` only in the context of actual writable scenarios, and we're not writing anything to this database.

I'm working on a better solution for all QA DB references, btw.

File: frontend/src/metabase-types/api/settings.ts [hunk 9]
```diff
@@ -285,6 +285,7 @@ export interface Settings {
   "user-visibility": string | null;
   "last-acknowledged-version": string | null;
   "show-static-embed-terms": boolean | null;
+  "last-used-database-id"?: number | null;
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase-types/api/settings.ts:288]: Great question @kamilmielnik !
This is something that I didn't know about, and it's not documented anywhere.

@ranquild explained it, and I plan to document it AND to update types to reflect this. But I thought it falls outside the scope of this PR.

tl;dr
`null` or `value` - user has access and it reads the actual value
`undefined` - user does not have permissions to fetch this

File: e2e/test/scenarios/native/native-database-source.cy.spec.js (hunk 3 above)
Comment (nemanjaglumac) [reply e2e/test/scenarios/native/native-database-source.cy.spec.js:203]: It is expensive, but it was the only reliable approach that gave me confidence this thing works.
I can explore some other options.

File: frontend/src/metabase-types/api/mocks/settings.ts [hunk 10]
```diff
@@ -227,5 +227,6 @@ export const createMockSettings = (
   "uploads-schema-name": null,
   "user-visibility": null,
   "last-acknowledged-version": "v1",
+  "last-used-database-id": 1,
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase-types/api/mocks/settings.ts:230]: I wouldn't do that. Technically these presents are not connected

File: frontend/src/metabase/query_builder/components/view/View.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/view/View.jsx:413]: We should `connect` early. The current QueryBuilder/View/ViewHeader structure is broken. This is closer to the correct way of doing it

File: e2e/test/scenarios/native/native-database-source.cy.spec.js (hunk 4 above)
Comment (ranquild) [reply e2e/test/scenarios/native/native-database-source.cy.spec.js:70]: +1

**Review by ranquild — COMMENTED** [review]
Looks good. I'm worried that the case with deleting the last selected database doesn't work well. There are some comments from @kamilmielnik about that. Could you double check?

File: test/metabase/models/user_test.clj [hunk 11]
```diff
@@ -194,12 +194,12 @@
...
-     (mt/with-temporary-setting-values [enable-password-login false]
-       (ldap.test/with-ldap-server
-         (invite-user-accept-and-check-inboxes! :invitor default-invitor , :accept-invite? false)
-         (is (seq (mt/regex-email-bodies #"/auth/login"))))))))
+  (testing "if sso enabled and password login is disabled, email should send a link to sso login"
+    (mt/with-premium-features #{:disable-password-login}
+      (mt/with-temporary-setting-values [enable-password-login false]
+        (ldap.test/with-ldap-server
+          (invite-user-accept-and-check-inboxes! :invitor default-invitor , :accept-invite? false)
+          (is (seq (mt/regex-email-bodies #"/auth/login"))))))))
```
Comment (oleggromov) [inline test/metabase/models/user_test.clj:202]: My 2c about changes like that — I am not sure which linters/prettifiers everyone uses for clojure, but probably different. Let's avoid changes like that because they unnecessarily change the history.

Comment (nemanjaglumac) [reply test/metabase/models/user_test.clj:202]: Oopsy.
I didn't even notice that because I have the whitespace turned off in GH UI.

Thanks

File: frontend/src/metabase-types/api/mocks/settings.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase-types/api/mocks/settings.ts:230]: E.g. leave 1

**Review by ranquild — APPROVED** [review]

File: e2e/test/scenarios/native/native-database-source.cy.spec.js (hunk 4 above)
Comment (nemanjaglumac) [reply e2e/test/scenarios/native/native-database-source.cy.spec.js:70]: Unfortunately, our caching logic creates all sorts of issues not just here, but in a lot of other places. Unless we do a hard refresh (which visiting a home page achieves in Cypress), most of this will not work.

This also applies to [the other comment](https://github.com/metabase/metabase/pull/38986#discussion_r1503801570) and [my answer](https://github.com/metabase/metabase/pull/38986#discussion_r1503871038) there.

File: e2e/test/scenarios/native/native-database-source.cy.spec.js (hunk 5 above)
Comment (nemanjaglumac) [reply e2e/test/scenarios/native/native-database-source.cy.spec.js:132]: Added a new test for this.
As with the other scenarios, unless there is a hard refresh user will still see the old database.

Spoke to AlexP about this and we concluded that this is a global problem we have - definitely not tied to just this PR.

**Review by calherries — COMMENTED** [review]
What happens when the last used database is deleted? Do we need a test for that?

Is there a FE benefit of implementing `last-used-native-database-id` as a user-local setting, instead of a new column of the `core_user` table? (fetched with `GET /api/user/current` and updated with `PUT /api/user/:id`)?

From my perpective on the BE, the advantage of adding a new column in the `core_user` table over using a setting is that it can be a foreign key to the database ID. Then we can do things like automatically set the value to NULL if the database is deleted.

**Review by calherries — COMMENTED** [review]
What happens when the last used database is deleted? Do we need a test for that?

Is there a FE benefit of implementing `last-used-native-database-id` as a user-local setting, instead of a new column of the `core_user` table? (fetched with `GET /api/user/current` and updated with `PUT /api/user/:id`)?

From my perpective on the BE, the advantage of adding a new column in the `core_user` table over using a setting is that it can be a foreign key to the database ID. Then we can do things like automatically set the value to NULL if the database is deleted.

**Comment by calherries** [comment]
Answering the bonus question here:

> Would it be ok to define a custom :getter for this new defsetting value that would first check if the database still exists AND if the current user has necessary permissions to query it?

Yes. We could also do this in `GET /api/user/current`

**Comment by nemanjaglumac** [comment]
> What happens when the last used database is deleted? Do we need a test for that?

@calherries 
We already have it [here](https://github.com/metabase/metabase/pull/38986/files#diff-7140ba4e9ce90fa912c83d5042756afd2c4ee3fb0369b0e078683027ca7b2742R59). Frontend will prompt the user to select a new database.

**Review by calherries — APPROVED** [review]
Looks good enough to merge for me like this anyway, it's more of a nit if you're handling this on the backend. Tests are tight too
